packages/core/src/routes/sign-in-experience/custom-ui-assets/index.ts

```diff
@@ -0,0 +1,120 @@
+  ...[router]: RouterInitArgs<T>
+) {
+  // TODO: Remove
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return;
+  }
+
+  router.post(
+    '/sign-in-exp/default/custom-ui-assets',
+    koaGuard({
+      files: object({
+        file: uploadFileGuard.array().min(1).max(1),
+      }),
+      response: z.object({
+        customUiAssetId: z.string(),
+      }),
+      status: [200, 400, 500],
+    }),
+    async (ctx, next) => {
+      const { file: bodyFiles } = ctx.guard.files;
+      const file = bodyFiles[0];
+
+      assertThat(file, 'guard.invalid_input');
+      assertThat(file.size <= maxUploadFileSize, 'guard.file_size_exceeded');
+      assertThat(file.mimetype === 'application/zip', 'guard.mime_type_not_allowed');
+
+      const { experienceZipsProviderConfig } = SystemContext.shared;
+      assertThat(
+        experienceZipsProviderConfig?.provider === 'AzureStorage',
+        'storage.not_configured'
+      );
+      const { connectionString, container } = experienceZipsProviderConfig;
+
+      const { uploadFile, downloadFile, isFileExisted } = buildAzureStorage(
+        connectionString,
+        container
+      );
+
+      const [tenantId] = await getTenantId(ctx.URL);
+      assertThat(tenantId, 'guard.can_not_get_tenant_id');
+
+      const customUiAssetId = generateStandardId(8);
+      const objectKey = `${tenantId}/${customUiAssetId}/assets.zip`;
+      const errorLogObjectKey = `${tenantId}/${customUiAssetId}/error.log`;
+
+      try {
+        // Upload the zip file to `experience-zips` container, in which a blob trigger is configured,
+        // and an azure function will be executed automatically to unzip the file on blob received.
+        // If the unzipping process succeeds, the zip file will be removed and assets will be stored in
+        // `experience-blobs` container. If it fails, the error message will be written to `error.log` file.
+        await uploadFile(await readFile(file.filepath), objectKey, {
+          contentType: file.mimetype,
+        });
+
+        const hasUnzipCompleted = async (retryTimes: number) => {
+          if (retryTimes > maxRetryCount) {
+            throw new AbortError('Unzip timeout. Max retry count reached.');
+          }
+          const [hasZip, hasError] = await Promise.all([
+            isFileExisted(objectKey),
+            isFileExisted(errorLogObjectKey),
+          ]);
+          if (hasZip) {
+            throw new Error('Unzip in progress...');
+          }
+          if (hasError) {
+            const errorLogBlob = await downloadFile(errorLogObjectKey);
+            const errorLog = await streamToString(errorLogBlob.readableStreamBody);
+            throw new AbortError(errorLog || 'Unzipping failed.');
+          }
+        };
+
+        await pRetry(hasUnzipCompleted, {
+          retries: maxRetryCount,
+          onFailedAttempt: (error) => {
+            if (error instanceof AbortError) {
+              throw error;
+            }
+          },
```

Comment:
i suspect the purpose of the `AbortError` is to skip this check. can you confirm?

Reply:
Let me check

Reply:
you're absolutely right. Using `AbortError` will automatically terminate the retry, so no need to handle it with `onFailedAttempt` any more.